src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs

```diff
@@ -1127,44 +1153,38 @@ private void ReadFromBuffer(Span<byte> buffer)
             return count;
         }
 
-        private async Task CopyFromBufferAsync(Stream destination, int count)
+        private async Task CopyFromBufferAsync(Stream destination, int count, CancellationToken cancellationToken)
         {
             Debug.Assert(count <= _readLength - _readOffset);
 
             if (NetEventSource.IsEnabled) Trace($"Copying {count} bytes to stream.");
-            await destination.WriteAsync(_readBuffer, _readOffset, count).ConfigureAwait(false);
+            await destination.WriteAsync(_readBuffer, _readOffset, count, cancellationToken).ConfigureAwait(false);
             _readOffset += count;
         }
 
-        private async Task CopyToAsync(Stream destination)
+        private Task CopyToUntilEofAsync(Stream destination, int bufferSize, CancellationToken cancellationToken)
         {
             Debug.Assert(destination != null);
 
             int remaining = _readLength - _readOffset;
-            if (remaining > 0)
-            {
-                await CopyFromBufferAsync(destination, remaining).ConfigureAwait(false);
-            }
+            return remaining > 0 ?
+                CopyToUntilEofWithExistingBufferedDataAsync(destination, cancellationToken) :
+                _stream.CopyToAsync(destination, bufferSize, cancellationToken);
```

Comment:
If _stream is NetworkStream, isn't this going to cause a buffer to be allocated?  Previously we were trying to use our own buffer here.  I can see how _stream.CopyToAsync is better for SslStream, but it seems like it's worse for NetworkStream, right?

Reply:
> isn't this going to cause a buffer to be allocated? 

NetworkStream will rent/return one from the ArrayPool.  Of course that could end up being an allocation, depending on what else is going on in the process.

> it seems like it's worse for NetworkStream, right?

Until/if we have a better solution in place, this will also avoid per-async receive Task allocations with NetworkStream.

That combined with it being better for SslStream seems to make this approach better.  But I can revert it if you disagree.

Reply:
Once SSLStream gets a CopyToAsync this might be better overall. You can’t avoid the allocation today though.

Reply:
Ok, I'm fine with it.

Reply:
#25659 seems to be the issue for tracking resettable awaitables.

If/when we add that, we could reconsider the approach here.  That said, (a) this only affects Connection: close responses (right?), which aren't super common; (b) it's probably better to use CopyToAsync for SslStream anyway.